app/models/manageiq/providers/openstack/cloud_manager.rb

```diff
@@ -40,6 +40,21 @@ def self.default_blacklisted_event_names
     )
   end
 
+  def hostname_uniqueness_valid?
+    return unless hostname_required?
+    return unless hostname.present? # Presence is checked elsewhere
+
+    existing_providers = Endpoint.where(:hostname => hostname.downcase).where.not(:resource_id => id)
+                                 .includes(:resource).select { |endpoint| endpoint.resource.uid_ems == uid_ems }
```

Comment:
Thanks for comments. Keystone_v3_domain_id alias is used now and where&select query is more readable than join query is, agreed.